tests/simulator.py

```diff
@@ -0,0 +1,253 @@
+    """Converts a level to a VCD understandable mnemonic."""
+    if level == Level.Z:
+        return "Z"
+    if level == Level.HIGH:
+        return "1"
+    return "0"
+
+
+@dataclasses.dataclass(frozen=True)
+class Change:
+    """Container to record simulation events."""
+
+    net_name: str
+    time_us: int
+    level: Level
+
+
+class Engine:
+    """Manages the overall simulation state and clock."""
+
+    def __init__(self) -> None:
+        self._start_us = int(time.monotonic() * 1e6)
+        self._nets: List["Net"] = []
+
+    def reset(self) -> None:
+        """Clears out all existing state and resets the simulation."""
+        self._start_us = int(time.monotonic() * 1e6)
+        self._nets = []
+
+    def _find_net_by_pin_id(self, pin_id: str) -> Optional["Net"]:
+        """Returns a net (if any) that has a pin with the given id."""
+        for net in self._nets:
+            if net.contains_pin_id(pin_id):
+                return net
+        return None
+
+    def create_net(
+        self, net_id: str, default_level: Level = Level.Z, monitor: bool = False
+    ) -> "Net":
+        """Creates a new net with the given name. Monitored nets are also traced."""
+        net = Net(net_id, default_level=default_level, monitor=monitor)
+        self._nets.append(net)
+        return net
+
+    def change_history(self) -> Sequence[Change]:
+        """Returns an ordered history of all events in monitored nets."""
+        monitored_nets = [net for net in self._nets if net.history]
+        combined: List[Change] = []
+        for net in monitored_nets:
+            if net.history:
+                for time_us, level in net.history:
+                    combined.append(Change(net.name, time_us, level))
+        combined.sort(key=lambda v: v.time_us)
+        return combined
+
+    def write_vcd(self, path: str) -> None:
+        """Writes monitored nets to the provided path as a VCD file."""
+        with open(path, "wt") as vcdfile:
+            vcdfile.write("$version pytest output $end\n")
+            vcdfile.write("$timescale 1 us $end\n")
+            vcdfile.write("$scope module top $end\n")
+            monitored_nets = [net for net in self._nets if net.history]
+            for net in monitored_nets:
+                vcdfile.write(f"$var wire 1 {net.name} {net.name} $end\n")
+            vcdfile.write("$upscope $end\n")
+            vcdfile.write("$enddefinitions $end\n")
+            combined = self.change_history()
+            # History starts when the engine is first reset or initialized.
+            vcdfile.write(f"#{self._start_us}\n")
+            last_us = self._start_us
+            for change in combined:
+                if change.time_us != last_us:
+                    vcdfile.write(f"#{change.time_us}\n")
+                    last_us = change.time_us
+                vcdfile.write(f"{_level_to_vcd(change.level)}{change.net_name}\n")
+
+
+# module global/singleton
+engine = Engine()
+
+
+class FakePin:
```

Comment:
I'd call it a DigitalInOut. We actually have a Pin class that does nothing currently. We've discussed enhancing it and calling it Pad though.

```suggestion
class FakeDigitalInOut:
```